frontend/src/libs/import-export/normalize-ddah-import.ts

```diff
@@ -0,0 +1,111 @@
+import { Applicant, MinimalDdah } from "../../api/defs/types";
+import {
+    SpreadsheetRowMapper,
+    matchByUtoridOrName,
+} from "../../libs/import-export";
+
+/**
+ * Convert imported spreadsheet or JSON data into an
+ * array of minimal DDAH objects.
+ *
+ * @param {({
+ *     fileType: "json" | "spreadsheet";
+ *     data: any;
+ * })} data
+ * @param {Applicant[]} applicants
+ * @returns {MinimalDdah[]}
+ */
+export function normalizeDdahImports(
+    data: {
+        fileType: "json" | "spreadsheet";
+        data: any;
+    },
+    applicants: Applicant[]
+): MinimalDdah[] {
+    let ret: MinimalDdah[] = [];
+
+    if (data.fileType === "json") {
+        let unwrapped: MinimalDdah[] = data.data;
+        if ((unwrapped as any).ddahs) {
+            unwrapped = (unwrapped as any).ddahs;
+        }
+        for (const ddah of unwrapped) {
+            ret.push(ddah);
+        }
+    }
+
+    if (data.fileType === "spreadsheet") {
+        const unwrapped = data.data;
+        // Get an upper bound for the maximum number of duties that the spreadsheet might have
+        const maxDuties = Math.round(
+            Math.max(
+                ...unwrapped.map((row: object) => Object.keys(row).length),
+                0
+            ) / 2
+        );
+
+        // We need to generate a keymap for all the likely column names
+        const keyMap: { [key: string]: string } = {
+            Position: "position_code",
+            "First Name": "first_name",
+            "Given Name": "first_name",
+            First: "first_name",
+            "Last Name": "last_name",
+            Surname: "last_name",
+            "Family Name": "last_name",
+            Last: "last_name",
+        };
+        // We will also add `Hours #` and `Duty #` to the keymap for the number of duties in our range
+        for (let i = 0; i <= maxDuties; i++) {
+            keyMap[`Duty ${i}`] = `duty_${i}`;
+            keyMap[`Hours ${i}`] = `hours_${i}`;
+            if (i < 10) {
+                keyMap[`Duty 0${i}`] = `duty_${i}`;
+                keyMap[`Hours 0${i}`] = `hours_${i}`;
+            }
+        }
+
+        // SpreadsheetRowMapper will perform fuzzy matching of column names for us.
+        const rowMapper = new SpreadsheetRowMapper({
+            keys: ["position_code", "first_name", "last_name", "utorid"],
+            keyMap,
+        });
+
+        for (const row of unwrapped) {
+            const normalized: {
+                [key: string]: any;
+            } = rowMapper.formatRow(row);
+            if (normalized.utorid == null) {
```

Comment:
just making sure we are including the case where `normalized.utorid` is `undefined` as well but excluding the case where it is an empty string.

Reply:
yes, that us correct. Only undefined values are converted to null. Empty strings are considered as valid values